lib/pacemaker/pcmk_sched_group.c

```diff
@@ -317,29 +317,42 @@ member_internal_constraints(gpointer data, gpointer user_data)
     member_data->previous_member = member;
 }
 
+/*!
+ * \internal
+ * \brief Create implicit constraints needed for a group resource
+ *
+ * \param[in,out] rsc  Group resource to create implicit constraints for
+ */
 void
-group_internal_constraints(pe_resource_t *rsc)
+pcmk__group_internal_constraints(pe_resource_t *rsc)
 {
     struct member_data member_data = { false, };
 
+    CRM_ASSERT(rsc != NULL);
+
+    /* Order group pseudo-actions relative to each other for restarting:
+     * stop group -> group is stopped -> start group -> group is started
+     */
+    pcmk__order_resource_actions(rsc, RSC_STOP, rsc, RSC_STOPPED,
+                                 pe_order_runnable_left);
     pcmk__order_resource_actions(rsc, RSC_STOPPED, rsc, RSC_START,
                                  pe_order_optional);
     pcmk__order_resource_actions(rsc, RSC_START, rsc, RSC_STARTED,
                                  pe_order_runnable_left);
-    pcmk__order_resource_actions(rsc, RSC_STOP, rsc, RSC_STOPPED,
-                                 pe_order_runnable_left);
 
     member_data.ordered = pe__group_flag_is_set(rsc, pe__group_ordered);
     member_data.colocated = pe__group_flag_is_set(rsc, pe__group_colocated);
     member_data.promotable = pcmk_is_set(uber_parent(rsc)->flags, pe_rsc_promotable);
     g_list_foreach(rsc->children, member_internal_constraints, &member_data);
 
     if (member_data.ordered && (member_data.previous_member != NULL)) {
+        // Stop group -> stop last member -> group is stopped
```

Comment:
I had to do a double-take because we say the same thing for each member in `member_internal_constraints()`, and here we do it again for the last member. The only difference is the flag -- "stop group" implies "stop last member."

It's not clear to me that we need to call the ordering function again with `pe_order_implies_then` for the last member. The last member's stop *doesn't* imply the second-to-last resource's stop and so on, so I don't see why we'd need to imply the last member's stop to trigger a domino effect. Rather, the group's first member has to stop before the group is "stopped", the second resource has to stop before the first can stop, and so on.

I have a shaky grasp here, but I commented out this entire block (`if (member_data.ordered...`), and all the scheduler regression tests still pass.

Reply:
You're right, I don't see any need for those. Added commit on top